src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -155,11 +131,13 @@ void CoreWorkerDirectActorTaskSubmitter::PushActorTask(
   rpc::Address addr(client.Addr());
   RAY_CHECK_OK(client.PushActorTask(
       std::move(request),
-      [this, addr, task_id](Status status, const rpc::PushTaskReply &reply) {
+      [this, addr, actor_id, task_id](Status status, const rpc::PushTaskReply &reply) {
         if (!status.ok()) {
-          task_finisher_->PendingTaskFailed(task_id, rpc::ErrorType::ACTOR_DIED, &status);
+          // Try to restart the task by killing the actor. This is not a great way to deal with
+          // failures, but a more robust solution would require extra work (see #6773).
+          KillActor(actor_id);
```

Comment:
We need to be a bit careful here because this callback could actually fire after the actor has already restarted.

Reply:
Yeah I'm definitely not happy with how this works right now, but I'm also not sure how to accurately determine the state of the actor right now without suffering from all those races. This may cause the actor to get killed unnecessarily (and also waste some reconstructions), but it will do the job of re-delivering the requests, so I guess it's still better than what happened before 🤷‍♂

A proper fix would probably involve an attempt to resend this particular message, but that's not very easy because the client is super stateful (e.g. it is entirely in charge of sequence numbers and doesn't expose them to anyone else, so right now there's no easy way to resend a message). Do you have any other suggesions?

Reply:
Long-term, it probably makes sense to version the messages and clients, so we should only kill the actor if its version matches the one sent with the task. For now, it's probably okay for this to be a bit inefficient.